### PR TITLE
Fix Top, Left, and Right padding for fullscreen android apps.

### PR DESCRIPTION
**Pull request by GaryQian**
The android WindowInsets does not take into account status and nav bars that are hidden. This change ensures that the top, left, and right edges have properly zeroed padding when it is necessary to correct for the hidden bars.

Top is zeroed when the status bar is hidden (called "Fullscreen" in android)

Left or Right or Both are zeroed when the navigation bar would appear on the respective sides when the device is in landscape mode.

This patch does not handle bottom padding when in portrait mode or landscape mode yet due to the issues with providing proper insets for on-screen keyboards.

### CODE REVIEW

**Comment by GaryQian** [comment]
Fixes: 
https://github.com/flutter/flutter/issues/18671
Related to:
https://github.com/flutter/flutter/issues/15947
https://github.com/flutter/flutter/issues/14432

**Comment by jonahwilliams** [comment]
@GaryQian I worked around some similar issues for accessibility here: https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L635

I would verify that these don't interact poorly - maybe we could ditch my hack?

**Comment by GaryQian** [comment]
I added a distinction between API < 23 always having navbar on right (if it is on the sides at all) and API >= 23 placing the navbar on the "bottom" of the device (left or right side depending on orientation)

I think as of now, the hacks are still necessary as the android API does not seem to provide clean ways to do it.

**Comment by GaryQian** [comment]
@cbracken @jason-simmons

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 1]
```diff
@@ -542,17 +542,62 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
         super.onSizeChanged(width, height, oldWidth, oldHeight);
     }
 
+    // Decide if we want to zero the sides to remove padding for hidden navbar.
```
Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Add a bit more detail. Specifically: 'zero the sides' -> 'zero the left/right-side padding of the window when the system navbar is hidden'.

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 2]
```diff
@@ -542,17 +542,62 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
...
+    enum ZeroSides { NONE, LEFT, RIGHT, BOTH }
+    ZeroSides calculateShouldZeroSides(WindowInsets insets) {
```
Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: I assume we'll need to update this code to handle the new cutout APIs. If that's the case, please add a TODO to help out the future implementor while you're here.

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 3]
```diff
@@ -542,17 +542,62 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
...
+                return ZeroSides.BOTH;
+            }
+        }
+        // Square orientation deprecated in API 16, we will not check for it and return false
+        // to be safe and not remove any unique padding for the devices that do use it.
+        return ZeroSides.NONE;
+    }
+
+    // This callback is not present in API < 20, which means lower API devices will see
+    // worse/improper padding as of now when the status and/or nav bar are hidden.
```
Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: `s/worse\///` and `s/as of now //` since it's unclear when 'now' is to the reader.

Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Could we preserve the existing behaviour for API < 20? We do [something along those lines](https://github.com/flutter/engine/blob/7ec1d211e4ae273d07f8cfde8442e103a915e914/shell/platform/darwin/ios/framework/Source/FlutterViewController.mm#L727-L734) in the iOS code.

Comment (GaryQian) [reply shell/platform/android/io/flutter/view/FlutterView.java]: The behavior for API < 20 is already what we do now.

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 4]
```diff
@@ -542,17 +542,62 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
...
+            (SYSTEM_UI_FLAG_HIDE_NAVIGATION & getWindowSystemUiVisibility()) == 0;
+
+        // We zero the left and/or right sides to prevent the padding the
+        // navigation bar would have caused.
+        ZeroSides zeroSides = ZeroSides.NONE;
+        if (!navigationBarVisible) {
+            zeroSides = calculateShouldZeroSides(insets);
+        }
+
+        mMetrics.physicalPaddingTop = !statusBarVisible ? 0 : insets.getSystemWindowInsetTop();
```
Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Uber-nitpick: I'd invert the condition to save readers one brain operation.

File: shell/platform/android/io/flutter/view/FlutterView.java [hunk 5]
```diff
@@ -542,17 +542,62 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
...
+            zeroSides == ZeroSides.RIGHT || zeroSides == ZeroSides.BOTH ? 0 : insets.getSystemWindowInsetRight();
         mMetrics.physicalPaddingBottom = 0;
-        mMetrics.physicalPaddingLeft = insets.getSystemWindowInsetLeft();
+        mMetrics.physicalPaddingLeft =
+            zeroSides == ZeroSides.LEFT || zeroSides == ZeroSides.BOTH ? 0 : insets.getSystemWindowInsetLeft();
 
         // Bottom system inset (keyboard) should adjust scrollable bottom edge (inset).
         mMetrics.physicalViewInsetTop = 0;
         mMetrics.physicalViewInsetRight = 0;
+        // TODO(garyq): Detect and distinguish between bottom nav padding and keyboard padding.
```
Comment (cbracken) [inline shell/platform/android/io/flutter/view/FlutterView.java]: Thanks for adding this. It's surprising to me that Android doesn't include specific APIs for getting the keyboard frame or even height (or perhaps I'm searching for the wrong things) :(

**Comment by cbracken** [comment]
LGTM on the logic. Just a couple minor nits/questions.

File: shell/platform/android/io/flutter/view/FlutterView.java (hunk 5 above)
Comment (GaryQian) [reply shell/platform/android/io/flutter/view/FlutterView.java]: It seems that the way most people do it is through heuristics. They compare the height of the padding to the height of the screen and over a certain amount is considered a keyboard.

**Pull request merged**